### PR TITLE
Code review pr

### PR DESCRIPTION
**Pull request by dan-da**
Addresses some issues brought up in code review:

1. removes use of the inaccurate word "Exception" in a comment.

2. removes use of word "callback" and replaces with "Fn f" in a comment, to match argument type.

3. re-implements tree::walk() to use a non-recursive algo.

4. (maybe controversial) calls panic!(), only in debug build, for case when op.timestamp equals timestamp of previous op.

**Item 4 merits more discussion/explanation**

Note:  I wrote the below before seeing PR #8 which implements alternative (b) below.  I think that is a fine approach, though possibly/probably we should use warn!() instead of info!().   I will merge these PRs in the morning.

<strike>This is a condition that breaks the "all ops must have unique timestamp" invariant/contract as specified in Section 3.1 of the paper.  quote: *timestamps 't need to be globally unique and totally ordered (e.g. Lamport Timestamps).*  In fact, in the crdt paper algo, they do not even check for this case.

The State impl in this crate does not try to strictly enforce the unique timestamp invariant because a) it would require keeping an index of op timestamps and b) it's not even fully possible when log truncation is enabled.  So we kind of need to rely on callers behaving correctly.

The panic!() is not really necessary as the op can be ignored.  It is put there only as means of alert during development.   Alternatives would be to:  a) remove debug panic and do nothing at all,  b) log error instead of panic,  c) return a Result<(),Error>.

(a) is a little scary because there's more chance of a uniqueness bug slipping by in calling code.  Finding it would likely require testing 2+ replicas with ops out of order and comparing their state.

(b) simple, and a nice compromise between (a) and (c).  I'm good with it.

(c) I don't really wish to return a result/error because the general idea is that all ops get applied, and it is weird for caller to have to check result of applying each op for a disallowed-by-contract edge case.  Nothing the caller can do about that case anyway except maybe log an error.  Plus, would need to re-work a bunch of test and example code.  However, it can be done if necessary.

anyway, that's my thinking so far.

### CODE REVIEW

File: src/state.rs [hunk 1]
```diff
@@ -174,15 +174,18 @@ impl<ID: TreeId, TM: TreeMeta, A: Actor> State<ID, TM, A> {
...
-                    // panic!("applying op with timestamp equal to previous op.  Every op should have a unique timestamp.");
+                    // We can panic to catch it during dev/test.
+                    #[cfg(debug_assertions)]
+                    panic!("applying op with timestamp equal to previous op.  Every op must have a unique timestamp.");
 
-                    // Production code should just treat it as a non-op.
-                    // #[cfg(not(debug_assertions))]
+                    // Production code can just treat it as a no-op.
+                    #[cfg(not(debug_assertions))]
+                    return;
```
Comment (dirvine) [inline src/state.rs:188]: I feel here we should do one thing and do it consistently. So if it is an error we need to at least log it or return an error. The latter is problematic though as rust-crdt trait requires no return type (no Result). So here I feel log is our best bet in both debug and release.

Comment (dan-da) [reply src/state.rs:188]: yes, I think that's good.  will make it so.

**Comment by dan-da** [comment]
rebased on top of #11.

File: examples/tree.rs [hunk 2]
```diff
@@ -107,8 +107,12 @@ impl<ID: TreeId, TM: TreeMeta, A: Actor + std::fmt::Debug> Replica<ID, TM, A> {
         // is the causally stable threshold.
 
         let mut v: Vec<&Clock<A>> = self.latest_time_by_replica.values().collect();
-        v.sort_unstable_by(|a, b| a.cmp(b));
-        v.pop() 
+        v.sort();
```
Comment (dirvine) [inline examples/tree.rs:110]: What do you think of 
`vec.sort();
vec.reverse():
vec.pop
`?

Comment (dan-da) [reply examples/tree.rs:110]: should work, but depending on Vec's implementation it could be hugely more expensive on a large list. So I'd prefer not to unless its shown/known that Vec's reverse() is fast.  I don't see any mention in the reverse() docs.

a couple notes:

1. VecDeque may be better suited for this, as it offers push_back and pop_front to use the start of list.
2. Unless I'm missing something the LogOpMove Vec could actually store at end instead of beginning.  I only did it this way because that's what they specify in the paper and I tried to follow their algo as exactly as I could.

I was actually rather surprised Vec doesn't have shift/unshift or equivalent for working with front of array.  Other langs/libs generally support this.

Comment (dan-da) [reply examples/tree.rs:110]: So I think this is faster on large lists:

```
        v.sort_by(|a, b| 
            match a.cmp(b) {
                Ordering::Greater => Ordering::Less,
                Ordering::Equal => Ordering::Equal,
                Ordering::Less => Ordering::Greater,
            }
        );
```

but it is also much uglier, and getting carried away.  :)

Really this is just sorting replicas, and it is just an example, so upon reflection I'm good with 'vec.sort(); vec.reverse(): vec.pop' for readability.

Comment (dan-da) [reply examples/tree.rs:110]: pushed.

**Review by dirvine — APPROVED** [review]
